File: LeetCode/101-symmetric-tree/ex_101.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        return root == NULL || checkSymmetric(root->left, root->right);
    }
    
private:
    bool checkSymmetric(TreeNode* left, TreeNode* right) {
        if (left == NULL && right == NULL)  return true;
        if (left == NULL || right == NULL)  return false;
        return left->val == right ->val && checkSymmetric(left->left, right->right) && checkSymmetric(right->left, left->right);
    }
};

File: LeetCode/101-symmetric-tree/ex_101.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean isSymmetric(TreeNode root) {
        return root == null || checkSymmetric(root.left, root.right);
    }
    
    private boolean checkSymmetric(TreeNode left, TreeNode right) {
        if (left == null && right == null)  return true;
        if (left == null || right == null)  return false;
        return left.val == right .val && checkSymmetric(left.left, right.right) && checkSymmetric(right.left, left.right);
    }
}